server/src/main/java/org/candlepin/util/X509CRLStreamWriter.java

```diff
@@ -0,0 +1,785 @@
+            DERBitString signatureBits = new DERBitString(signature);
+            out.write(signatureBits.getDEREncoded());
+        }
+        catch (DataLengthException e) {
+            throw new IOException("Could not sign", e);
+        }
+        catch (CryptoException e) {
+            throw new IOException("Could not sign", e);
+        }
+    }
+
+    @SuppressWarnings("rawtypes")
+    protected byte[] incrementCrlNumber(byte[] extensions) throws IOException {
+        DERTaggedObject taggedExts = (DERTaggedObject) DERTaggedObject.fromByteArray(extensions);
+        DERSequence seq = (DERSequence) taggedExts.getObject();
+        ASN1EncodableVector modifiedExts = new ASN1EncodableVector();
+
+        // Now we need to read the extensions and find the CRL number and increment it,
+        // and determine if its length changed.
+        Enumeration objs = seq.getObjects();
+        while (objs.hasMoreElements()) {
+            DERSequence ext = (DERSequence) objs.nextElement();
+            DERObjectIdentifier oid = (DERObjectIdentifier) ext.getObjectAt(0);
+            if (X509Extension.cRLNumber.equals(oid)) {
+                X509Extension newNumberExt =
+                    new X509Extension(false, new DEROctetString(newCrlNumber.getDEREncoded()));
+
+                ASN1EncodableVector crlNumber = new ASN1EncodableVector();
+                crlNumber.add(X509Extension.cRLNumber);
+                crlNumber.add(newNumberExt.getValue());
+                modifiedExts.add(new DERSequence(crlNumber));
+            }
+            else {
+                modifiedExts.add(ext);
+            }
+        }
+
+        DERSequence seqOut = new DERSequence(modifiedExts);
+        DERTaggedObject out = new DERTaggedObject(true, 0, seqOut);
+        return out.getDEREncoded();
+    }
+
+    protected int handleHeader(OutputStream out) throws IOException {
+        int addedEntriesLength = 0;
+        for (DERSequence s : newEntries) {
+            addedEntriesLength += s.getDEREncoded().length;
+        }
+
+        int topTag = readTag(crlIn, null);
+        int topTagNo = readTagNumber(crlIn, topTag, null);
+        int oldTotalLength = readLength(crlIn, null);
+
+        // Now we are in the TBSCertList
+        int tbsTag = readTag(crlIn, null);
+        int tbsTagNo = readTagNumber(crlIn, tbsTag, null);
+        int oldTbsLength = readLength(crlIn, null);
+
+        /* We may need to adjust the overall length of the tbsCertList
+         * based on changes in the revokedCertificates sequence, so we
+         * will cache the tbsCertList data in this temporary byte stream.
+         */
+        ByteArrayOutputStream temp = new ByteArrayOutputStream();
+
+        int tagNo = DERTags.NULL;
+        Date oldThisUpdate = null;
+        while (true) {
+            int tag = readTag(crlIn, null);
+            tagNo = readTagNumber(crlIn, tag, null);
+
+            if (tagNo == GENERALIZED_TIME || tagNo == UTC_TIME) {
```

Comment:
> Are we running the risk of getting stuck here? Is there a way for this to miss and not be true?

At some point, we have to hit a GENERALIZED_TIME or UTC_TIME that represents the `thisUpdate` component.  Not hitting one of those would mean the CRL is not compliant with the RFC and corrupt.  Regardless, the code would not run forever because eventually the stream would be emptied by all the read operations and an exception would get thrown.